packages/mui-lab/src/Masonry/Masonry.js

```diff
@@ -182,80 +182,95 @@ const Masonry = React.forwardRef(function Masonry(inProps, ref) {
-        const childHeight = parseToNumber(childComputedStyle.height)
-          ? Math.ceil(parseToNumber(childComputedStyle.height)) + childMarginTop + childMarginBottom
-          : 0;
-        if (childHeight === 0) {
-          skip = true;
-          return;
-        }
-        // if there is a nested image that isn't rendered yet, masonry's height shouldn't be computed yet
-        for (let i = 0; i < child.childNodes.length; i += 1) {
-          const nestedChild = child.childNodes[i];
-          if (nestedChild.tagName === 'IMG' && nestedChild.clientHeight === 0) {
-            skip = true;
-            break;
+          if (parentWidth === 0 || childWidth === 0) {
+            return;
           }
-        }
-        if (!skip) {
-          // find the current shortest column (where the current item will be placed)
-          const currentMinColumnIndex = columnHeights.indexOf(Math.min(...columnHeights));
-          columnHeights[currentMinColumnIndex] += childHeight;
-          const order = currentMinColumnIndex + 1;
-          child.style.order = order;
-        }
-      });
-      if (!skip) {
-        setMaxColumnHeight(Math.max(...columnHeights));
-        const numOfLineBreaks = currentNumberOfColumns > 0 ? currentNumberOfColumns - 1 : 0;
-        setNumberOfLineBreaks(numOfLineBreaks);
-      }
-    };
 
+          const currentNumberOfColumns = Math.round(
+            parentWidth / (childWidth + firstChildMarginLeft + firstChildMarginRight),
+          );
+
+          const columnHeights = new Array(currentNumberOfColumns).fill(0);
+          let skip = false;
+          masonry.target.childNodes.forEach((child) => {
+            if (
+              child.nodeType !== Node.ELEMENT_NODE ||
+              child.dataset.class === 'line-break' ||
+              skip
+            ) {
+              return;
+            }
+            const childComputedStyle = window.getComputedStyle(child);
+            const childMarginTop = parseToNumber(childComputedStyle.marginTop);
+            const childMarginBottom = parseToNumber(childComputedStyle.marginBottom);
+            // if any one of children isn't rendered yet, masonry's height shouldn't be computed yet
+            const childHeight = parseToNumber(childComputedStyle.height)
+              ? Math.ceil(parseToNumber(childComputedStyle.height)) +
+                childMarginTop +
+                childMarginBottom
+              : 0;
+            if (childHeight === 0) {
+              skip = true;
+              return;
+            }
+            // if there is a nested image that isn't rendered yet, masonry's height shouldn't be computed yet
+            for (let i = 0; i < child.childNodes.length; i += 1) {
+              const nestedChild = child.childNodes[i];
+              if (nestedChild.tagName === 'IMG' && nestedChild.clientHeight === 0) {
+                observer.current.observe(nestedChild);
```

Comment:
@mnajdova 

Image masonry was broken because `elements[1]` inside `ResizeObserver` handler can be `undefined` as ResizeObserver only outputs elements that have resized. So, if masonry container resized but first child didn't, `elements[1]` is undefined, leading to not doing further computation. I suggest the following. What do you think?

```diff
diff --git a/packages/mui-lab/src/Masonry/Masonry.js b/packages/mui-lab/src/Masonry/Masonry.js
index 46cf39d4d3..e8c9b09bc4 100644
--- a/packages/mui-lab/src/Masonry/Masonry.js
+++ b/packages/mui-lab/src/Masonry/Masonry.js
@@ -187,14 +187,16 @@ const Masonry = React.forwardRef(function Masonry(inProps, ref) {
       ? undefined
       : new ResizeObserver((elements) => {
           const masonry = elements[0];
-          const masonryFirstChild = elements[1];
+          const masonryFirstChild = elements[1] || masonryRef.current.firstChild;
           if (!masonryFirstChild) {
             return;
           }
 
           const parentWidth = masonry.contentRect.width;
-          const childWidth = masonryFirstChild.contentRect.width;
-          const firstChildComputedStyle = window.getComputedStyle(masonryFirstChild.target);
+          const childWidth = elements[1]
+            ? masonryFirstChild.contentRect.width
+            : masonryFirstChild?.clientWidth || 0;
+          const firstChildComputedStyle = window.getComputedStyle(masonryFirstChild.target || masonryFirstChild);
           const firstChildMarginLeft = parseToNumber(firstChildComputedStyle.marginLeft);
           const firstChildMarginRight = parseToNumber(firstChildComputedStyle.marginRight);
 
@@ -233,7 +235,6 @@ const Masonry = React.forwardRef(function Masonry(inProps, ref) {
             for (let i = 0; i < child.childNodes.length; i += 1) {
               const nestedChild = child.childNodes[i];
               if (nestedChild.tagName === 'IMG' && nestedChild.clientHeight === 0) {
-                observer.current.observe(nestedChild);
                 skip = true;
                 break;
               }
```

Reply:
Does this mean that the indexes are not really stable? For example we cannot guarantee that `elements[0]` is always the masonry.

Reply:
Yep, I addressed that here!: https://github.com/mui-org/material-ui/pull/29452/commits/256bbfc6282dcedcf40e054403ca1730a273a2fc